health/health.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package health
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type HealthReport struct {
+	Live  bool
+	Ready bool
+}
+
+type reporterState struct {
+	// The health indicators that this reporter reports.
+	reports HealthReport
+
+	// Expiry time for this reporter's reports.
+	timeout time.Duration
+
+	// The most recent report.
+	latest HealthReport
+
+	// Time of that most recent report.
+	timestamp time.Time
+}
+
+type HealthState struct {
+	// Mutex to protect concurrent access to this health state.
+	mutex *sync.Mutex
+
+	// Map from reporter name to corresponding state.
+	reporters map[string]*reporterState
+}
+
+func (state *HealthState) RegisterReporter(
+	// Unique name for this reporter.
```

Comment:
This comments-interlaced-with-params approach is certainly readable but does it work with godoc etc?